### PR TITLE
[ASNetworkImageNode] fix for contentMode when using resizable defaultImage

### PR DESCRIPTION
**Pull request by tomizimobile**
This is a proposed fix for #1362. It tracks changes in `contentMode` when a resizable image is set for `defaultImage`, and applies the pending content mode before setting the image.

### CODE REVIEW

**Comment by appleguy** [comment]
@tomizimobile this seems interesting - I don't quite understand it in detail, though.  Could you add some additional comments, even just a description walkthrough here?

Does this handle cases where the network node loads fully, but then its URL is set to nil, and it needs to revert to the defaultImage?  There are many such paths...

**Comment by tomizimobile** [comment]
@appleguy A walkthrough: as mentioned in the original issue, the layer's `contentsGravity` is set to resize if an image is stretchable, and at no other point is it set to the desired `contentMode`. This code checks if the `defaultImage` is stretchable when set, and if so, will track changes in the `contentMode`. The `_pendingContentModeState` is also set at that time to the current `contentMode` in case it was set before `defaultImage` was. Then, in `-setImage:`, if the image isn't the defaultImage, it'll set the `contentMode` to the pending state first. My flag `_needsToApplyPendingContentMode` ensures the comparison of the image to the `defaultImage` and setting `contentMode` are only done when necessary.

I hadn't considered any case where the image is set to the `defaultImage` after the node loads. I'm pretty sure I'll need to make some changes - I'll take a closer look at it and add some comments in the source when I have a chance.

**Comment by tomizimobile** [comment]
@appleguy I reworked the code a bit, but it doesn't seem to work well with animated images. I have a test project that I was using to test this with that is very basic - just showing an ASNetworkImageNode with a test image url - but animated images only seem to work correctly a small portion of the time (or if there's a delay in setting the url). I can't track down exactly why but maybe someone else can give it a look?

**Comment by hannahmbanana** [comment]
@tomizimobile: could you share your test project? thanks!

**Comment by tomizimobile** [comment]
@hannahmbanana Here's my test project - I've defined `pngURL` and `gifURL` for testing with, and you can see where I set the image URL at line 86 of PostViewController.swift.

[asdk-issue-1822.zip](https://github.com/facebook/AsyncDisplayKit/files/418046/asdk-issue-1822.zip)

**Comment by keleftheriou** [comment]
Can one of the admins verify this patch?

**Comment by appleguy** [comment]
test this please

**Comment by appleguy** [comment]
^^ command to trigger jenkins to see if tests pass

**Comment by maicki** [comment]
Hey @tomizimobile, could you please rebase your current branch on top of master so the CI can pass? That would be awesome, Thanks!

**Comment by tomizimobile** [comment]
@maicki I've updated the code - I'm still seeing an issue with animated images though. Not sure what I'm missing with those, maybe someone else has input on that.

**Comment by maicki** [comment]
@tomizimobile Thanks for updating! Can we see it in your sample project you attached?

**Comment by tomizimobile** [comment]
@maicki whoops, forgot to attach a modified project, it's been updated for Swift 3 and new layout APIs. You can see lines 87-89 of PostViewController.swift show where i set the URL. With the example gif, you can see it almost always gets stretched still, although curiously doesn't if a delay is applied. The fix seems ok for normal images though.

[asdk-issue-1822.zip](https://github.com/facebook/AsyncDisplayKit/files/604820/asdk-issue-1822.zip)

**Comment by garrettmoon** [comment]
@tomizimobile If you would be willing to get the project building again I promise I'll actually take a look at this (ASDK in the podfile is currently pointing to a repo on your desktop and it doesn't build when pointed at your personal master).

**Comment by tomizimobile** [comment]
@garrettmoon hmmm, that project worked fine for me when pointed at my master. Regardless, I've updated the project and merged my master with the current master, and it seems to be working as expected on the first run. For some reason though, now it doesn't want to load the gif at all on subsequent runs - don't really have the time at the moment to dig into it further to see if it's something I've added or a different bug, but here's the project: [asdk-issue-1822.zip](https://github.com/facebook/AsyncDisplayKit/files/764913/asdk-issue-1822.zip)

**Comment by garrettmoon** [comment]
@tomizimobile GIFs are actually broken in ASDK at the moment :( But fixed if you point to PINRemoteImage master.

**Comment by garrettmoon** [comment]
I'll be trying to get the PINRemoteImage GIF fix into ASDK ASAP. We can circle back on this then.

**Comment by Adlai-Holler** [comment]
Jenkins add to whitelist and test this please.

File: AsyncDisplayKit/ASNetworkImageNode.mm [hunk 1]
```diff
@@ -121,6 +126,17 @@ - (void)setImage:(UIImage *)image
...
+  // this will ensure the contentMode is properly set when the final image is loaded
+  if (_hasStretchableDefaultImage) {
+    if (!_needsToApplyPendingContentMode && ASObjectIsEqual(image, _defaultImage)) {
+      _pendingContentModeState = self.contentMode;
+      _needsToApplyPendingContentMode = YES;
+    } else if (_needsToApplyPendingContentMode && !ASObjectIsEqual(image, _defaultImage)) {
+      _needsToApplyPendingContentMode = NO;
+      self.contentMode = _pendingContentModeState;
+    }
+  }
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/ASNetworkImageNode.mm]: Minor – we should guard the new flags with `__instanceLock__`. Unfortunately that will mean either:

- Holding the lock while calling into super (who will probably reacquire the lock)
- Relinquishing the lock before calling into super (and breaking the atomicity of these operations)

I don't really like either of those options, but I think option #1 is the best for the time being. When it comes to recursive locks, "in for a penny, in for a pound" 😈

File: AsyncDisplayKit/ASNetworkImageNode.mm [hunk 2]
```diff
@@ -193,6 +210,23 @@ - (UIImage *)defaultImage
   return _defaultImage;
 }
 
+- (void)setAnimatedImage:(id<ASAnimatedImageProtocol>)animatedImage
+{
+  if (_needsToApplyPendingContentMode) {
```
Comment (garrettmoon) [inline AsyncDisplayKit/ASNetworkImageNode.mm]: As @Adlai-Holler said, these two methods need to be guarded with the instance lock.

**Comment by garrettmoon** [comment]
Hey @tomizimobile we're going to close this for now as it's getting stale, but we'll gladly re-open if you address comments and rebase. Thanks for putting this up!

**Pull request closed without merging**